.circleci/config.yml

```diff
@@ -0,0 +1,116 @@
+          command: |
+            echo 'export PATH=$HOME/.local/cmake/bin:$HOME/.cargo/bin:$HOME/.local/bin:$PATH' >> $BASH_ENV
+            echo 'export LD_RUN_PATH=$LD_RUN_PATH:$HOME/.local/lib' >> $BASH_ENV
+            echo 'export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$HOME/.local/lib' >> $BASH_ENV
+            echo 'export LDFLAGS=-L$HOME/.local/lib' >> $BASH_ENV
+            echo 'export CPPFLAGS=-I$HOME/.local/include' >> $BASH_ENV
+            echo 'export CXXFLAGS=-I$HOME/.local/include' >> $BASH_ENV
+            echo 'export LOCAL_PREFIX=$HOME/.local' >> $BASH_ENV
+            echo 'export LD_RUN_PATH=$LD_RUN_PATH:$HOME/.local/lib' >> $BASH_ENV
+            echo 'export LD_RUN_PATH=$LD_RUN_PATH:$HOME/.local/lib' >> $BASH_ENV
+            source $BASH_ENV
+      - checkout
+      - run: 
+          name: Installing Build Dependencies
+          command: |
+            sudo add-apt-repository ppa:ubuntu-toolchain-r/test -y;
+            sudo apt-get update -y;
+            sudo apt-get install -y g++-4.8 curl libdw-dev libelf-dev elfutils binutils-dev
+      - restore_cache:
+          name: Restoring ~/.local Cache
+          keys: 
+            - v1-local-{{ .Branch }}-{{ .Environment.CIRCLE_SHA1 }}
+            - v1-local-{{ .Branch }}
+            - v1-local
+      - restore_cache:
+          name: Restoring ~/.cargo Cache
+          keys:
+            - v1-cargo-{{ .Branch }}-{{ checksum "Cargo.lock" }}
+            - v1-cargo-{{ .Branch }}
+            - v1-cargo
+
+      - run: 
+          name: Setting up Rust and the rest of the Build Environment
+          command: |
+            RUST_VERSION=`tail -n 1 RUST_VERSION`
+            if [[ ! -e $HOME/.cargo ]]; then
+              curl https://sh.rustup.rs -sSf |
+                sh -s -- --no-modify-path --default-toolchain ${RUST_VERSION} -y;
+              rustup default ${RUST_VERSION};
+            else
+              rustup default ${RUST_VERSION};
+            fi
+
+            make -f ci-build/Makefile prepare-rustfmt
+
+            if [[ ! -f $HOME/.local/cmake/bin/cmake ]]; then
+              curl https://cmake.org/files/v3.10/cmake-3.10.0-Linux-x86_64.tar.gz -sSf | tar xzf -;
+              rm -rf $HOME/.local/cmake
+              mv cmake-3.10.0-Linux-x86_64 $HOME/.local/cmake
+            fi
+
+            if [[ ! -e $HOME/.local/lib/libgflags.a ]]; then
+              cd /tmp && \
+              curl -L https://github.com/gflags/gflags/archive/v2.1.2.tar.gz -o gflags.tar.gz && \
+              tar xf gflags.tar.gz && \
+              cd gflags-2.1.2 && \
+              cmake -DCMAKE_INSTALL_PREFIX=${LOCAL_PREFIX} . && \
+              make -j && \
+              make install;
+            fi
+      # check format first
+      -  run:
+          name: Checking Format
+          command: make format && git diff-index --quiet HEAD -- || (git diff; echo please make format and run tests before creating a PR!; exit 1)
+      - restore_cache:
```

Comment:
Does it not work to restore the `target` before running the format?

Reply:
Yes, it works but restoring the `./target` and maybe `sccache` for the future may take a long time to process (about five minutes), checking format first may reduce feedback time for the PR authors.

Reply:
That's a great reason. :+1: